### PR TITLE
feature/create profile page & edit profile details

### PR DESCRIPTION
**Pull request by Nandan-18**
Change Description:
- Create a on-device user object that contains user details
- Create a profile page that fetches user details and displays it
- Fetch profile picture from Firebase Storage using stored on-device URI
- Edit Profile Page allows user to delete and modify all details, including profile picture

### CODE REVIEW

**Review by Sundance636 — APPROVED** [review]
Looks good!

File: app/src/main/java/com/example/scanpal/EditProfileFragment.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+import com.bumptech.glide.Glide;
+import com.google.android.material.floatingactionbutton.FloatingActionButton;
+import com.google.android.material.textfield.TextInputEditText;
+import com.google.android.material.textfield.TextInputLayout;
+import com.google.firebase.firestore.FirebaseFirestore;
+
+import java.util.Objects;
+
+public class EditProfileFragment extends Fragment {
```
Comment (Logan9312) [inline app/src/main/java/com/example/scanpal/EditProfileFragment.java:30]: Docstring

File: app/src/main/java/com/example/scanpal/EditProfileFragment.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        saveButton.setOnClickListener(v -> saveUserDetails());
+        goBack.setOnClickListener(v -> {
+            NavController navController = NavHostFragment.findNavController(EditProfileFragment.this);
+            navController.navigate(R.id.save_profile_edits);
+        });
+
+        fetchUserDetails();
+    }
+
+    private void openGallery() {
```
Comment (Logan9312) [inline app/src/main/java/com/example/scanpal/EditProfileFragment.java:88]: Docstring

File: app/src/main/java/com/example/scanpal/EditProfileFragment.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        Intent intent = new Intent(Intent.ACTION_PICK);
+        intent.setType("image/*");
+        pickImageLauncher.launch(intent);
+    }
+
+    private void uploadImageToFirebase(Uri imageUri) {
```
Comment (Logan9312) [inline app/src/main/java/com/example/scanpal/EditProfileFragment.java:94]: Docstring

File: app/src/main/java/com/example/scanpal/EditProfileFragment.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        imageController.uploadImage(imageUri,
+                taskSnapshot -> Toast.makeText(getContext(), "Image Uploaded Successfully", Toast.LENGTH_SHORT).show(),
+                e -> Toast.makeText(getContext(), "Upload failed: " + e.getMessage(), Toast.LENGTH_SHORT).show());
+    }
+
+    private void fetchUserDetails() {
```
Comment (Logan9312) [inline app/src/main/java/com/example/scanpal/EditProfileFragment.java:100]: Docstring

File: app/src/main/java/com/example/scanpal/EditProfileFragment.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+                @Override
+                public void onError(Exception e) {
+                    Toast.makeText(getContext(), "Failed to fetch user details: " + e.getMessage(), Toast.LENGTH_LONG).show();
+                }
+            });
+        }
+    }
+
+    private void saveUserDetails() {
```
Comment (Logan9312) [inline app/src/main/java/com/example/scanpal/EditProfileFragment.java:122]: Docstring

File: app/src/main/java/com/example/scanpal/ImageController.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+import android.net.Uri;
+
+import com.google.android.gms.tasks.OnFailureListener;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.firebase.storage.FirebaseStorage;
+import com.google.firebase.storage.StorageReference;
+import com.google.firebase.storage.UploadTask;
+
+public class ImageController {
```
Comment (Logan9312) [inline app/src/main/java/com/example/scanpal/ImageController.java:11]: Docstring

File: app/src/main/java/com/example/scanpal/ProfileFragment.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+import androidx.annotation.Nullable;
+import androidx.fragment.app.Fragment;
+import androidx.navigation.NavController;
+import androidx.navigation.fragment.NavHostFragment;
+
+import com.bumptech.glide.Glide;
+import com.google.android.material.floatingactionbutton.FloatingActionButton;
+import com.google.firebase.firestore.FirebaseFirestore;
+
+public class ProfileFragment extends Fragment {
```
Comment (Logan9312) [inline app/src/main/java/com/example/scanpal/ProfileFragment.java:21]: Docstring

File: app/src/main/java/com/example/scanpal/ProfileFragment.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        super.onResume();
+        String username = userController.fetchStoredUsername();
+        if (username != null) {
+            fetchUserDetails(username);
+        } else {
+            Toast.makeText(getContext(), "No username found in internal storage", Toast.LENGTH_SHORT).show();
+        }
+    }
+
+    private void initializeViews(View view) {
```
Comment (Logan9312) [inline app/src/main/java/com/example/scanpal/ProfileFragment.java:64]: Docstring

File: app/src/main/java/com/example/scanpal/ProfileFragment.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        buttonYourEvents = view.findViewById(R.id.button_your_events);
+        buttonHomepage = view.findViewById(R.id.button_homepage);
+        buttonEditProfile = view.findViewById(R.id.button_edit_profile);
+        addUsername = view.findViewById(R.id.addUsername);
+        firstName = view.findViewById(R.id.first_name);
+        lastName = view.findViewById(R.id.last_name);
+        homepage = view.findViewById(R.id.homepage);
+    }
+
+    private void setupButtonListeners() {
```
Comment (Logan9312) [inline app/src/main/java/com/example/scanpal/ProfileFragment.java:79]: Docstring

File: app/src/main/java/com/example/scanpal/ProfileFragment.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+            // TODO
+        });
+
+        buttonEditProfile.setOnClickListener(v -> {
+            NavController navController = NavHostFragment.findNavController(ProfileFragment.this);
+            navController.navigate(R.id.edit_profile);
+        });
+    }
+
+    private void fetchUserDetails(String username) {
```
Comment (Logan9312) [inline app/src/main/java/com/example/scanpal/ProfileFragment.java:110]: Docstring

File: app/src/main/java/com/example/scanpal/UserController.java [hunk 11]
```diff
@@ -55,6 +78,31 @@ public void addUser(User user, UserAddCallback callback) {
         });
     }
 
+    public void updateUser(User user, UserUpdateCallback callback) {
```
Comment (Logan9312) [inline app/src/main/java/com/example/scanpal/UserController.java:81]: Docstring

File: app/src/main/java/com/example/scanpal/UserController.java [hunk 12]
```diff
@@ -100,4 +162,18 @@ public void isUsernameTaken(String username, UsernameCheckCallback callback) {
 
     }
 
-}
+    // fetch username from internal storage
+    public String fetchStoredUsername() {
```
Comment (Logan9312) [inline app/src/main/java/com/example/scanpal/UserController.java:166]: Should be changed to a docstring for consistency with other method comments

File: app/src/main/java/com/example/scanpal/UserUpdateCallback.java [hunk 13]
```diff
@@ -0,0 +1,7 @@
+package com.example.scanpal;
+
+public interface UserUpdateCallback {
```
Comment (Logan9312) [inline app/src/main/java/com/example/scanpal/UserUpdateCallback.java:3]: Docstring

**Review by Logan9312 — CHANGES_REQUESTED** [review]
Added comments to make Samuel happy

**Review by Logan9312 — APPROVED** [review]

**Pull request merged**